### PR TITLE
[DX-1150] Updated request size limit

### PR DESCRIPTION
**Pull request by andyo-tyk**
### Preview Link
https://deploy-preview-4198--tyk-docs.netlify.app/docs/nightly/basic-config-and-security/control-limit-traffic/request-size-limits/

### Description
Updated Request Size Limit documentation, adding Tyk OAS and aligning structure with other 5.3.0 changes

___

## **Type**
enhancement, documentation

___

## **Description**
- Enhanced the Request Size Limits documentation to include broader security and middleware focus, detailed use cases, and clarification on how request size limiting works.
- Added new documentation for configuring Request Size Limit middleware with both Tyk Classic and Tyk OAS APIs, including example configurations.
- Updated the documentation menu structure to reflect the new documentation pages for Tyk OAS and Tyk Classic API configurations.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>request-size-limits.md</strong><dd><code>Enhanced Request Size Limits Documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md
<li>Updated tags and introduction to reflect broader security and <br>middleware focus.<br> <li> Added detailed use cases for the Request Size Limit middleware.<br> <li> Clarified how request size limiting works, including precedence and <br>error responses.<br> <li> Provided instructions for applying a system level size limit.<br> <li> Included references to configuring request size limits with Tyk OAS <br>and Tyk Classic APIs.

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4198/files#diff-63bea101e4e6fb9151159c22227b536fc3e1cade9c052a33681491205f9c10cb">+44/-57</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>request-size-limit-tyk-classic.md</strong><dd><code>New Documentation for Request Size Limit with Tyk Classic APIs</code></dd></summary>
<hr>

tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic.md
<li>Introduced documentation for configuring Request Size Limit middleware <br>with Tyk Classic APIs.<br> <li> Detailed how to apply size limits at different levels: system-wide, <br>API-level, and endpoint-level.<br> <li> Provided example configurations for each level of granularity.

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4198/files#diff-166d68392a6265f9b5850342586cf69b3f7eee873f261d2bcc7508a2425c8e51">+78/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>request-size-limit-tyk-oas.md</strong><dd><code>New Documentation for Request Size Limit with Tyk OAS APIs</code></dd></summary>
<hr>

tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md
<li>Introduced documentation for configuring Request Size Limit middleware <br>with Tyk OAS APIs.<br> <li> Explained the absence of API-level rate limit for Tyk OAS and provided <br>a workaround.<br> <li> Provided example configurations for endpoint-level request size <br>limits.

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4198/files#diff-3565a648321c9e7a85ed1386d9e12a0ccbe1bfcca5583ca1948db312f329ec72">+93/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>menu.yaml</strong><dd><code>Updated Menu Structure for Request Size Limits Documentation</code></dd></summary>
<hr>

tyk-docs/data/menu.yaml
<li>Updated the menu structure to include separate entries for Tyk OAS and <br>Tyk Classic configurations under Request Size Limits.

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk-docs/pull/4198/files#diff-a6789c66bfa0660f7fdbe4796002ab4777677bae59a0276689c840100ff0b0b2">+14/-2</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md [hunk 1]
```diff
@@ -1,83 +1,70 @@
...
-You can optionally configure a request size limit (in bytes) to an API by adding `global_size_limit` to the `version` element of the API Definition, for example:
-```
-"global_size_limit": 2500 
-```
+#### Limiting request sizes for a lightweight microservice
+A specific API is designed to handle lightweight, fast transactions and is not equipped to process large payloads. You can set an API-level size limit that ensures the microservice behind this API is not forced to handle requests larger than it is designed for, maintaining its performance and efficiency.
 
-This limit is applied for all endpoints _within an API_. It is evaluated after the Gateway-wide size limit and before any endpoint-specific size limit. If this test fails, the Tyk Gateway will report `400 Request is too large`.
+#### Controlling the size of GraphQL queries
+A GraphQL API endpoint might be susceptible to complex queries that can lead to performance issues. By setting a request size limit for the GraphQL endpoint, you ensure that overly complex queries are blocked, protecting the backend services from potential abuse and ensuring a smooth operation.
```
Comment (dcs3spp) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md:22]: For GraphQL does max_query_depth also improve performance?

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md [hunk 2]
```diff
@@ -1,83 +1,70 @@
...
-    
-    {{< img src="/img/2.10/request_size_limit.png" alt="Plugins drop down" >}}
+<hr>
 
-4.  Set the size limit in bytes.
-    
-    {{< img src="/img/2.10/request_size_settings.png" alt="Size limit form" >}}
+### Applying a system level size limit
+You can configure a request size limit (in bytes) that will be applied to all APIs on your Tyk Gateway by adding `max_request_body_size` to the `http_server_options` [element]({{< ref "tyk-oss-gateway/configuration#http_server_optionsmax_request_body_size" >}}) of your `tyk.conf` Gateway configuration. For example:
+```
```
Comment (dcs3spp) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md]: For JSON fragments in this PR, does``` yaml render ok?

```suggestion
```yaml
```

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md [hunk 3]
```diff
@@ -1,83 +1,70 @@
...
+Tyk Cloud Classic enforces a strict request size limit of 1MB on all inbound requests via our cloud architecture. This limit does not apply to Tyk Cloud users.
+{{< /note >}}
 
+<hr>
 
+If you're using Tyk OAS APIs, then you can find details and examples of how to configure an API or endpoint-level request size limit [here]({{< ref "product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas" >}}).
 
+If you're using Tyk Classic APIs, then you can find details and examples of how to configure an API or endpoint-level request size limit [here]({{< ref "product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic" >}}).
 
+<!-- proposed "summary box" to be shown graphically on each middleware page
```
Comment (dcs3spp) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md:68]: Remember to revisit this later

File: tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
+---
+title: Using the Request Size Limit middleware with Tyk Classic APIs
+date: 2024-03-04
+description: "Using the Request Size Limit middleware with Tyk Classic APIs"
+tags: ["request size limit", "size limit", "security", "middleware", "per-endpoint", "per-API", "Tyk Classic", "Tyk Classic API"]
+---
+
+The [request size limit]({{< ref "basic-config-and-security/control-limit-traffic/request-size-limits" >}}) middleware enables you to apply limits to the size of requests made to your HTTP APIs. You might use this feature to protect your Tyk Gateway or upstream services from avoid excessive memory usage or brute force attacks.
```
Comment (dcs3spp) [inline tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic.md]:
```suggestion
The [request size limit]({{< ref "basic-config-and-security/control-limit-traffic/request-size-limits" >}}) middleware enables you to apply limits to the size of requests made to your HTTP APIs. You might use this feature to protect your Tyk Gateway or upstream services from excessive memory usage or brute force attacks.
```

File: tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+If you're using the newer Tyk OAS APIs, then check out the [Tyk OAS]({{< ref "product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas" >}}) page.
+
+## Configuring the middleware in the Tyk Classic API Definition
+There are three different levels of granularity that can be used when configuring a request size limit.
+ - [system-wide]({{< ref "basic-config-and-security/control-limit-traffic/request-size-limits#applying-a-system-wide-size-limit" >}}): affecting all APIs deployed on the gateway
+ - [API-level]({{< ref "product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic#applying-a-size-limit-for-a-specific-api" >}}): affecting all endpoints for an API
+ - [endpoint-level]({{< ref "product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic#applying-a-size-limit-for-a-specific-endpoint" >}}): affecting a single API endpoint
+
+### Applying a size limit for a specific API
+You can configure a request size limit (in bytes) to an API by configuring the `global_size_limit` to the `version` element of the API Definition, for example:
```
Comment (dcs3spp) [inline tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic.md]:
```suggestion
You can configure a request size limit (in bytes) to an API by configuring the `global_size_limit` within the `version` element of the API Definition, for example:
```

File: tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                "path": "/anything",
+                "method": "POST",
+                "size_limit": 100
+            }
+        ]
+    }
+}
+```
+
+In this example the endpoint-level Request Size Limit middleware has been configured for HTTP `POST` requests to the `/anything` endpoint. For any call made to this endpoint, Tyk will check the size of the payload (Request body) and, if it is larger than 100 bytes, will reject the request, returning `HTTP 400 Request is too large`.
```
Comment (dcs3spp) [inline tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-classic.md:59]:
```suggestion
In this example the endpoint-level Request Size Limit middleware has been configured for HTTP `POST` requests to the `/anything` endpoint. For any call made to this endpoint, Tyk will check the size of the request body and, if it is larger than 100 bytes, will reject the request, returning `HTTP 400 Request is too large`.
```

File: tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+In this example the endpoint-level Request Size Limit middleware has been configured for HTTP `POST` requests to the `/anything` endpoint. For any call made to this endpoint, Tyk will check the size of the payload (Request body) and, if it is larger than 100 bytes, will reject the request, returning `HTTP 400 Request is too large`.
+
+The configuration above is a complete and valid Tyk OAS API Definition that you can import into Tyk to try out the virtual endpoint middleware.
+
+## Configuring the middleware in the API Designer
+Adding a Virtual Endpoint to your API endpoints is easy when using the API Designer in the Tyk Dashboard, simply follow the steps taken in this short video:
+
+ < placeholder for video >
```
Comment (dcs3spp) [inline tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md]: remember to add placeholder video

**Review by dcs3spp — CHANGES_REQUESTED** [review]
PR LGTM, made suggestions and requires peer review

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md [hunk 8]
```diff
@@ -1,83 +1,70 @@
...
+#### Protecting the entire Tyk Gateway from DDoS attacks
+You can configure a system-wide request size limit that protects all APIs managed by the Tyk Gateway from being overwhelmed by excessively large requests, which could be part of a DDoS attack, ensuring the stability and availability of the gateway.
 
-### Applying a size limit for a specific API
-You can optionally configure a request size limit (in bytes) to an API by adding `global_size_limit` to the `version` element of the API Definition, for example:
-```
-"global_size_limit": 2500 
-```
+#### Limiting request sizes for a lightweight microservice
+A specific API is designed to handle lightweight, fast transactions and is not equipped to process large payloads. You can set an API-level size limit that ensures the microservice behind this API is not forced to handle requests larger than it is designed for, maintaining its performance and efficiency.
```
Comment (andrei-tyk) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md]: wouldn't "and is generally not equipped to" so we don't sound categorical?

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md [hunk 9]
```diff
@@ -1,83 +1,70 @@
...
 
-This limit is applied for all endpoints _within an API_. It is evaluated after the Gateway-wide size limit and before any endpoint-specific size limit. If this test fails, the Tyk Gateway will report `400 Request is too large`.
+#### Controlling the size of GraphQL queries
+A GraphQL API endpoint might be susceptible to complex queries that can lead to performance issues. By setting a request size limit for the GraphQL endpoint, you ensure that overly complex queries are blocked, protecting the backend services from potential abuse and ensuring a smooth operation.
 
-### Applying a size limit for a specific API endpoint
-Tyk provides a _Request Size Limit_ middleware that can be configured per API endpoint. This gives you the most granular control over request sizes.
+#### Restricting upload size on a file upload endpoint
+An API endpoint is designed to accept file uploads, but to prevent abuse, you want to limit the size of uploads to 1MB. You can configure the Request Size Limit middleware for this endpoint with a limit of 1MB. This prevents users from uploading excessively large files, protecting your storage and bandwidth resources.
```
Comment (andrei-tyk) [inline tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md]: "You can configure the Request Size Limit middleware for this endpoint with a limit of 1MB." tells you about what the user can do but is not linked with the preceding or the next statement in terms of reasoning. Wouldn't it be better if the sentence would start with "To achieve this, you can configure the Request..."

File: tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
+---
+title: Using the Request Size Limit middleware with Tyk OAS APIs
+date: 2024-03-04
+description: "Using the Request Size Limit middleware with Tyk OAS APIs"
+tags: ["request size limit", "size limit", "security", "middleware", "per-endpoint", "per-API", "Tyk OAS", "Tyk OAS API"]
+---
+
+The [request size limit]({{< ref "basic-config-and-security/control-limit-traffic/request-size-limits" >}}) middleware enables you to apply limits to the size of requests made to your HTTP APIs. You might use this feature to protect your Tyk Gateway or upstream services from avoid excessive memory usage or brute force attacks.
```
Comment (andrei-tyk) [inline tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md]: "avoid" seems out of place here too

File: tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+ - [API-level]({{< ref "product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas#applying-a-size-limit-for-a-specific-api" >}}): affecting all endpoints for an API
+ - [endpoint-level]({{< ref "product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas#applying-a-size-limit-for-a-specific-endpoint" >}}): affecting a single API endpoint
+
+### Applying a size limit for a specific API
+The API-level rate limit has not yet been implemented for Tyk OAS APIs.
+
+You can work around this by implementing a combination of endpoint-level rate limits and [allow]({{< ref "advanced-configuration/transform-traffic/endpoint-designer#allowlist" >}}) or [block]({{< ref "advanced-configuration/transform-traffic/endpoint-designer#blocklist" >}}) lists.
+
+### Applying a size limit for a specific endpoint
+The design of the Tyk OAS API Definition takes advantage of the `operationID` defined in the OpenAPI Document that declares both the path and method for which the middleware should be added.
```
Comment (andrei-tyk) [inline tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md]: I think it's "operationId" (at least that's what I found in the code )

**Comment by dcs3spp** [comment]
@andyo-tyk I've resolved most comments. Unresolved comments relate to question concerning [max_query_depth](https://github.com/TykTechnologies/tyk-docs/pull/4198#discussion_r1514146324), reminders re placeholder video and yaml formatting for JSON fragment

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md (hunk 1 above)
Comment (andyo-tyk) [reply tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md:22]: I have to admit, I have no idea. I think that's a potential enhancement to these docs to be added later.

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md (hunk 2 above)
Comment (andyo-tyk) [reply tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md]: image

Looks OK to me - I'll add this.

File: tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md (hunk 8 above)
Comment (andyo-tyk) [reply tyk-docs/content/basic-config-and-security/control-limit-traffic/request-size-limits.md]: This is meant to be "in the scenario where you have an API that is not equipped to process large payloads"... so I'll tweak the wording.

File: tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md (hunk 10 above)
Comment (andyo-tyk) [reply tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md]: Yes, that's a typo - good spot. Thanks.

File: tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md (hunk 11 above)
Comment (andyo-tyk) [reply tyk-docs/content/product-stack/tyk-gateway/middleware/request-size-limit-tyk-oas.md]: Argh.... that's in all of the middleware pages... good spot... I'll correct it here and then sweep up the others elsewhere. I note that I got it right in the following paragraph and in the code example 🙈

**Review by dcs3spp — APPROVED** [review]
@andyo-tyk PR LGTM

@andrei-tyk would you be able to peer review and once approved will merge and release

**Review by andrei-tyk — APPROVED** [review]

**Pull request merged**